src/analysis/processing/qgsalgorithmcalculateoverlaps.cpp

```diff
@@ -0,0 +1,199 @@
+      mOutputFields.append( QgsField( QStringLiteral( "%1_pc" ).arg( vl->name() ), QVariant::Double ) );
+    }
+  }
+
+  mOutputType = mSource->wkbType();
+  mCrs = mSource->sourceCrs();
+  mInputCount = mSource->featureCount();
+  mInputFeatures = mSource->getFeatures();
+  return true;
+}
+
+QVariantMap QgsCalculateVectorOverlapsAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  QString destId;
+  std::unique_ptr< QgsFeatureSink > sink( parameterAsSink( parameters, QStringLiteral( "OUTPUT" ), context, destId, mOutputFields,
+                                          mOutputType, mCrs ) );
+  if ( !sink )
+    throw QgsProcessingException( invalidSinkError( parameters, QStringLiteral( "OUTPUT" ) ) );
+
+  // build a spatial index for each constraint layer for speed. We also store input constraint geometries here,
+  // to avoid refetching and projecting them later
+  QMap< QString, QgsSpatialIndex > spatialIndices;
+  for ( auto it = mOverlayerSources.begin(); it != mOverlayerSources.end(); ++ it )
+  {
+    feedback->pushInfo( QObject::tr( "Preparing %1" ).arg( it->first ) );
+    QgsFeatureIterator featureIt = it->second->getFeatures( QgsFeatureRequest().setSubsetOfAttributes( QgsAttributeList() ).setDestinationCrs( mCrs, context.transformContext() ).setInvalidGeometryCheck( context.invalidGeometryCheck() ).setInvalidGeometryCallback( context.invalidGeometryCallback() ) );
+    spatialIndices.insert( it->first, QgsSpatialIndex( featureIt, feedback, QgsSpatialIndex::FlagStoreFeatureGeometries ) );
+  }
+
+  // loop through input
+  double step = mInputCount > 0 ? 100.0 / mInputCount : 0;
+  long i = 0;
+  QgsFeature feature;
+  while ( mInputFeatures.nextFeature( feature ) )
+  {
+    if ( feedback->isCanceled() )
+      break;
+
+    QgsAttributes outAttributes = feature.attributes();
+    if ( feature.hasGeometry() )
+    {
+      const QgsGeometry inputGeom = feature.geometry();
+
+      // prepare for lots of intersection tests (for speed)
+      std::unique_ptr< QgsGeometryEngine > bufferGeomEngine( QgsGeometry::createGeometryEngine( inputGeom.constGet() ) );
+      bufferGeomEngine->prepareGeometry();
+
+      // calculate overlap attributes
+
+      for ( auto it = mOverlayerSources.begin(); it != mOverlayerSources.end(); ++ it )
+      {
+        if ( feedback->isCanceled() )
+          break;
+
+        QgsSpatialIndex &index = spatialIndices[ it->first ];
+        const QList<QgsFeatureId> matches = index.intersects( inputGeom.boundingBox() );
+        QVector< QgsGeometry > intersectingGeoms;
+        intersectingGeoms.reserve( matches.count() );
+        for ( QgsFeatureId match : matches )
+        {
+          const QgsGeometry overlayGeometry = index.geometry( match );
+          if ( bufferGeomEngine->intersects( overlayGeometry.constGet() ) )
+          {
+            intersectingGeoms.append( overlayGeometry );
+          }
```

Comment:
Just in case a lot of matches
```suggestion
          }
          
          if ( feedback->isCanceled() )
￼	          break;
```